src/Raven.Server/Documents/Sharding/Handlers/Batches/ShardedBatchCommand.cs

```diff
@@ -86,17 +89,40 @@ public IEnumerator<SingleShardedCommand> GetEnumerator()
                 commandStream = bufferedCommand.ModifyIdentityStream(cmd);
             }
 
-            int shardNumber = GetShardNumberForCommandType(cmd, bufferedCommand.IsServerSideIdentity);
+            var cmdResult = GetShardNumberForCommandType(cmd, bufferedCommand.IsServerSideIdentity);
             var stream = cmd.Type == CommandType.AttachmentPUT ? AttachmentStreams[streamPosition++] : null;
 
+            AssertBehavior(behavior, cmd.Type, ref previousBucket, cmdResult.Bucket);
+
             yield return new SingleShardedCommand
             {
-                ShardNumber = shardNumber,
+                ShardNumber = cmdResult.ShardNumber,
                 AttachmentStream = stream,
                 CommandStream = commandStream,
                 PositionInResponse = positionInResponse++
             };
         }
+
+        static void AssertBehavior(ShardedBatchBehavior behavior, CommandType commandType, ref int? previousBucket, int bucket)
+        {
+            if (previousBucket == null)
+            {
+                previousBucket = bucket;
+                return;
+            }
+
+            switch (behavior)
+            {
+                case ShardedBatchBehavior.TransactionalSingleBucketOnly:
+                    if (previousBucket != bucket)
+                        throw new ShardedBatchBehaviorViolationException($"Batch command of type '{commandType}' operates on a shard bucket '{bucket}' which violates the requested sharded batch behavior to operate on a single bucket ('{previousBucket}').");
```

Comment:
Here, need to add the document id of the document for this and the previous one. Because the user cannot usually tell the `bucket` to the document id that it belongs to

Reply:
New message:
```
Batch command of type 'PUT' operates on a document ID 'orders/1' with shard bucket '96772' which violates the requested sharded batch behavior to operate on a single bucket ('649571' from document ID 'companies/1').
```

Reply:
That is great message

Reply:
Maybe suggest cluster wide transaction here, too?

Reply:
```
Batch command of type 'PUT' operates on a document ID 'orders/1' with shard bucket '96772' which violates the requested sharded batch behavior to operate on a single bucket ('649571' from document ID 'companies/1'). Consider using cluster transactions in order to change documents from multiple buckets without violating transaction boundaries.
```

Reply:
Added `Consider using cluster transactions in order to modify documents from multiple buckets without violating transaction boundaries.` at the end of the message.